Add Search Application telemetry to xpack/usage

sloanelybutsurely opened the pull request:
implements xpack info and xpack usage for enterprise search. right now xpack usage only returns a count of search applications.

Commit:
search application telemetry via xpack.usage

sloanelybutsurely commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java:
```diff
@@ -101,6 +102,7 @@ public class SearchApplicationIndexService {
     public final NamedWriteableRegistry namedWriteableRegistry;
     private final BigArrays bigArrays;
 
+    @Inject
     public SearchApplicationIndexService(
         Client client,
         ClusterService clusterService,
```
@afoucret hello! kathleen suggested i reach out to you and ask if you know of a better way of making this `SearchApplicationIndexService` injectable. we looked at adding an instance in the return value of [`createComponents`](https://github.com/sloanelybutsurely/elasticsearch/blob/search-applications-telemetry/x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearch.java#L150) in `EnterpriseSearch.java` but noticed we didn't have a `BigArrays` instance available.

do you know how we could get a `BigArrays` instance or perhaps some better way?

kderusso replied:
Potentially `BigArrays.NON_RECYCLING_INSTANCE`? But I'm not familiar enough with this part of our code to confidently say that's a good suggestion.

carlosdelest replied:
I remember @jimczi advocating for this service to be kept as non-injectable as it is a thin abstraction that wasn't worth to extract as a service (we started having that as a service and it was changed). I'm struggling to find that conversation though 😓 

 Is there any change to the usage of the service to revert that decision back? (disclaimer: I would be a huge fan of doing it 😅 )

kderusso replied:
If we're using the service to provide telemetry information that may be enough of a justification - however, if we figure out the correct way to add BigArrays we can create it in `EnterpriseSearch` and return it so it's accessible where it needs to be. I don't have super strong feelings about it other than the hesitation to make the service injectable without hearing @jimczi 's feedback.

jimczi replied:
We don't really need to call the service directly? Can we rely on the action instead, something like:
```
 client.execute(ListSearchApplicationAction.INSTANCE, ...
```

Commit:
impl EnterpriseSearchInfoTransportAction, fix test

Commit:
fix docs yaml tests and security operator tests

kderusso commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/application/EnterpriseSearchFeatureSetUsageTests.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class EnterpriseSearchFeatureSetUsageTests extends AbstractWireSerializingTestCase<EnterpriseSearchFeatureSetUsage> {
```
Nitpick: I think we might want to rename this to `EnterpriseSearchFeatureSetUsageSerializingTests` - that matches the name we've given to other serialization tests and also keeps `EnterpriseSeearchFeatureSetUsageTests` free for any future tests we may want to add to the usage object itself.

sloanelybutsurely replied:
✅ https://github.com/elastic/elasticsearch/pull/95306/commits/85e0913b3f32b86bfd852698e24ea646ee28bbac

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/100_usage.yml:
```diff
@@ -0,0 +1,40 @@
+        ignore: 404
+
+  - do:
+      indices.delete:
```
May want to ignore: 404 here too

kderusso commented on x-pack/plugin/ent-search/qa/rest/build.gradle:
```diff
@@ -7,7 +7,7 @@ dependencies {
 restResources {
   restApi {
-    include '_common', 'cluster', 'nodes', 'indices', 'index', 'search_application'
+    include '_common', 'cluster', 'nodes', 'indices', 'index', 'search_application', 'xpack.usage'
```
Is this necessary? It doesn't seem to be included in other module gradle builds.

sloanelybutsurely replied:
if i remove it this is what i get:

> java.lang.RuntimeException: Failure at [entsearch/100_usage:30]: Rest api [xpack.usage] cannot be found in the rest spec. Either it doesn't exist or is missing from the test classpath. Check the 'restResources' block of your project's build.gradle file.

kderusso replied:
Thanks for verifying!

kderusso commented on x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/100_usage.yml:
```diff
@@ -0,0 +1,40 @@
+setup:
```
This test looks great! 

Some other test cases you could consider: 
- count = 0
- Add a new search application and verify the count increases / remove and verify it decreases

kderusso reviewed (commented):
Wow, @sloanelybutsurely , great job for your first Elasticsearch PR! 

I've got a few comments, but nothing huge. 

As for the failing test, it looks like it's failing on V7 REST Compatibility (`./gradlew :rest-api-spec:yamlRestTestV7CompatTest`). I've kicked this off locally but it takes a while to run, I'll keep you posted if I see a failure. 

Since this is a new addition to telemetry, the test may have to be [muted](https://github.com/elastic/elasticsearch/blob/ac9edce20c25492866f570897154c56b1677fa06/REST_API_COMPATIBILITY.md#troubleshooting-compatibility-test-failures) for V7.

kderusso commented:
Update: V7 compatibility tests also ran locally for me, but it could be worth trying muting it to fix the build.

carlosdelest commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/application/EnterpriseSearchFeatureSetUsage.java:
```diff
@@ -0,0 +1,69 @@
+public class EnterpriseSearchFeatureSetUsage extends XPackFeatureSet.Usage {
+
+    public static final String SEARCH_APPLICATIONS = "search_applications";
+    private final Map<String, Object> searchApplicationsStats;
```
Shouldn't we use a specific class for the `SearchApplicationStats`? That would allow us to understand what stats are gathered, and that's what other examples like [RemoteClusterFeatureSetUsage](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/RemoteClusterFeatureSetUsage.java) do...

I know for now it's just a little information, but it would be easier to extend that way...

ioanatia replied:
this should be fine for what we are doing right now, we can always add this later, it's not required for this PR IMO.
also we can find other examples that just use a `Map<String, Object>` and they seem to work fine:

https://github.com/elastic/elasticsearch/blob/ba69ff661078ab0a65240f08199f28c95a22c2b6/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MachineLearningFeatureSetUsage.java#L34-L37

carlosdelest replied:
Fine with me then - I'm such a fan of strong typing! If you're all ok I'm as well 🙂

carlosdelest commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchUsageTransportAction.java:
```diff
@@ -0,0 +1,66 @@
+        searchApplicationIndexService.listSearchApplication("*", 0, 0, listener.delegateFailure((l, searchApplicationResult) -> {
+            l.onResponse(
+                new XPackUsageFeatureResponse(
+                    new EnterpriseSearchFeatureSetUsage(true, true, Map.of("count", searchApplicationResult.totalResults()))
```
Is Enterprise Search always available and enabled? Shouldn't we check it like we did on `EnterpriseSearchInfoTransportAction`?

ioanatia replied:
looking at other modules and the response of `_xpack/usage` is depending on the license level,
`enabled` should be `XPackSettings.ENTERPRISE_SEARCH_ENABLED.get(settings)` and `available` should be `LicenseUtils.LICENSED_ENT_SEARCH_FEATURE.checkWithoutTracking(licenseState)`.

example: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherUsageTransportAction.java

sloanelybutsurely replied:
✅ https://github.com/elastic/elasticsearch/pull/95306/commits/fb485618ba87f9e9c538dda60326e55109e96b24

carlosdelest reviewed (commented):
Superb work. Not an easy one for your first PR as it involved many moving parts. Thanks for showing how it's done! 💯 

I have a comment on creating a specific class for the usage stats - I think we should do that. The rest is spotless IMHO.

carlosdelest commented:
> Update: V7 compatibility tests also ran locally for me, but it could be worth trying muting it to fix the build.

@sloanelybutsurely , have you tried bringing this branch up to date with `main`? Some compat tests break because of that, would be worth a try 👍

Commit:
set enabled & available with settings & license

Commit:
test count increases and decreases

Commit:
include xpack instead of xpack.usage specifically

Commit:
rename enterprise search usage serializing tests

carlosdelest reviewed (approved):
💯

kderusso reviewed (approved):
Approving, but I'd feel more comfortable waiting until Monday to check with @jimczi on the open dependency injection question before merging. WDYT?

jimczi commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchUsageTransportAction.java:
```diff
@@ -0,0 +1,80 @@
+        ClusterState state,
+        ActionListener<XPackUsageFeatureResponse> listener
+    ) throws Exception {
+        searchApplicationIndexService.listSearchApplication("*", 0, 0, listener.delegateFailure((l, searchApplicationResult) -> {
```
I think we should call the action instead. See my comment below.

jimczi replied:
Should we also handle the case where the module is not enabled?
When not enabled the request should return an empty response?
With the current implementation the call will fail since the endpoint does not exist?

jimczi reviewed (commented):
The logic looks good to me. I left some comments regarding the usage of the index service.

Commit:
call action directly

Co-authored-by: Kathleen DeRusso <kathleen.derusso@elastic.co>

Commit:
handle `enabled = false` for entsearch usage

sloanelybutsurely commented:
@jimczi i've updated this to call the action directly instead of having a dependency on the service. i've also added a case to handle enterprise search being disable. ptal?

Commit:
return empty stats on error

jimczi commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchUsageTransportAction.java:
```diff
@@ -0,0 +1,109 @@
+                enabled,
+                Collections.emptyMap()
+            );
+            listener.onResponse(new XPackUsageFeatureResponse(usage));
```
Silently ignoring exceptions like this seems wrong. I [see](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportXPackInfoAction.java#L83) that simply throwing the exception could break the entire info action call so that's not an option either. 
It's also a bigger scope than just adding a new info endpoint for ent-search so it's fine to keep it this way in this PR. 
I just wanted to raise the awareness that this is not something we should do in general ;) 
I'll open a more generic issue to discuss error handling in the `XPackUsageFeatureTransportAction`.

sloanelybutsurely replied:
for what it's worth this is a pattern used elsewhere: https://github.com/elastic/elasticsearch/blob/7fb344238dcad615282b70557b70c67eede482ca/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearningUsageTransportAction.java#L131-L150

jimczi commented on x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchUsageTransportAction.java:
```diff
@@ -0,0 +1,109 @@
+            indexNameExpressionResolver
+        );
+        this.licenseState = licenseState;
+        this.client = client;
```
Calling the action (list search application) directly requires to have the `manage_search_application` privilege. 
We can authorise this specific action to always have access to the search application internals by using the [`ent search origin`](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ClientHelper.java#L193):
```
this.client = new OriginSettingClient(client, ENT_SEARCH_ORIGIN);
```
The rest tests are working in this PR because the test user has the [`manage_search_application` privilege](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/ent-search/qa/rest/roles.yml#L3). Using the ent-search origin will ensure that this privilege is no longer required.

jimczi reviewed (commented):
Thanks @sloanelybutsurely. The change looks good. I left one last comment that requires a small change (the ent-search origin),

Commit:
use an OriginSettingClient

jimczi reviewed (approved):
LGTM. 
A small comment on methodology, we don't use force merge usually when working on a PR. You can simply merge main in your branch so that you don't lose the commit history and reviewers can look at latest commits when iterating on a PR.
